### PR TITLE
detect default format name on output file extension

### LINKED ISSUE #1569: have render(...) infer output_format when output_file specified
By filing an issue to this repo, I promise that

- [X] I have fully read the issue guide at https://yihui.name/issue/.
- [X] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [X] I have learned the Github Markdown syntax, and formatted my issue correctly.

---

This is a suggestion, to solve a minor annoyance for me and maybe help some other people.

I mostly use R from the command line, rather than via Rstudio. I am working on integrating some rmarkdown report generation in a `make`-based approach:

```{makefile}
report-%.pdf: report.Rmd # ... some other dependencies set by % wildcard
	R -e "rmarkdown::render('$<', output_file='$@', output_format='pdf_document')" --args $(filter-out $<,$^)
```

Translating a bit for those less familiar with make: this specifies how to a create a target file named report-X.pdf (or report-Y.pdf, report-something.pdf, etc) from report.Rmd and other dependencies. `$<` resolves to report.Rmd and `$@` to the name of the target.  I was slightly surprised to discover that I had to manually specify `output_format=...` for this work.  When unspecified, I get the error:

```
> rmarkdown::render('report.Rmd', output_file='report-test.pdf')

processing file: report.Rmd
  |.................................................................| 100%
  ordinary text without R code

output file: report.knit.md

/usr/bin/pandoc +RTS -K512m -RTS report.utf8.md --to html4 --from markdown+autolink_bare_uris+ascii_identifiers+tex_math_single_backslash+smart --outputreport-test.pdf --email-obfuscation none --self-contained --standalone --section-divs --template /home/carl/libs/Rpackages/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable 'theme:bootstrap' --include-in-header /tmp/RtmppWfeua/rmarkdown-str602f78928780.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --metadata pagetitle=report.utf8.md
cannot produce pdf output from html4
Error: pandoc document conversion failed with error 1
Execution halted
make: *** [Makefile:32: report-test.pdf] Error 1
```

Obviously, I have the solution to avoid that error.  However, I'd like my make rule to be:

```{makefile}
report-%.pdf report-%.html report-%.docx: report.Rmd # ... some other dependencies set by % wildcard
	R -e "rmarkdown::render('$<', output_file='$@')" --args $(filter-out $<,$^)
```

This would allow the user to build whichever of the desired target types, and keep generation logic concise, consistent, etc. That approach would maintain the inference logic within the library, where details like `docx==word` are more readily known and minimize future external changes if there are tweaks to internal rmarkdown definitions.

Seems like something akin to

```{r}
# ...
if (missing(output_format) & !missing(output_file)) {
  output_format <- sprintf("%s_document", gsub("doc.*","word",gsub("^.+\\.(pdf|doc.*|html)$","\\1",output_file)))
}
# ...
```

in the `render` logic could do the trick, but I haven't dug into the details to see what other conflicts there might be, etc.

### PR DESCRIPTION
**Pull request by cderv**
This tries to offer a way of having other default than `html_document` when no format output is provided. 

This closes #1569. 

Idea is to at least use the `output_file` extension to offer 3 base default format : `html_document`, `pdf_document`, `docx_document`.

### CODE REVIEW

**Comment by cderv** [comment]
First idea in first commit was to guess the default from `render()` but it was not a good idea as we need to rely on a default behavior of taking the first format found in the YAML if none is provided in `render()`. 

The default format to use if none has been found is now based on the extension to improve a bit if `.pdf` or `.docx` is used in `output_file`

**Comment by cderv** [comment]
🤔 this create an issue I did not see coming. I need to look into it...

**Comment by cderv** [comment]
As I changed implementation, I forgot some things. It is fixed now.

File: R/output_format.R [hunk 1]
```diff
@@ -597,6 +601,17 @@ create_output_format <- function(name,
...
 }
 
+output_format_string_from_ext <- function(output_file) {
+  default_format <- "html_document"
+  if (is.null(output_file)) return(default_format)
+  switch(xfun::file_ext(output_file),
+    html = "html_document",
+    pdf = "pdf_document",
+    docx = "word_document",
+    "html_document" # always been the default format in R Markdown
```
Comment (yihui) [inline R/output_format.R:611]: Instead of hard-coding the format:

```suggestion
    default_format  # always been the default format in R Markdown
```

**Review by yihui — APPROVED** [review]
Looks good to me. Thanks!

**Pull request merged**